cachi2/core/package_managers/yarn/project.py

```diff
@@ -0,0 +1,129 @@
+"""
+Parse the relevant files of a yarn project.
+
+I also provides basic utility functions. The main logic to resolve and prefetch
+the dependencies should be implemented in other modules.
+"""
+import json
+import logging
+import re
+from typing import Any, NamedTuple, Optional, TypedDict
+
+from yaml import safe_load
+
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+_YARNRC_VERSION_REGEX = re.compile("yarn-[0-9].[0-9].[0-9].cjs")
+
+
+class Plugin(TypedDict):
+    """A plugin defined in a yarnrc.yml file.
+
+    Example:
+    - path: .yarn/plugins/@yarnpkg/plugin-exec.cjs
+      spec: "@yarnpkg/plugin-exec"
+    """
+
+    path: str
+    spec: str
+
+
+class YarnRc:
+    """A yarnrc file.
+
+    This class abstracts the underlying attributes and only exposes what
+    is relevant for the request processing.
+    """
+
+    def __init__(self, path: RootedPath, data: dict[str, Any]) -> None:
+        """Initiaze a YarnRc object."""
+        self._path = path
+        self._data = data
+
+    @property
+    def plugins(self) -> list[Plugin]:
+        """A list of plugins."""
+        return self._data.get("plugins", [])
+
+    @property
+    def yarn_path(self) -> Optional[str]:
+        """Path to the yarn script present in this directory."""
+        return self._data.get("yarnPath", None)
+
+    @property
+    def yarn_version(self) -> Optional[str]:
+        """Yarn version used in this project.
+
+        Extracted from the contents of the yarnPath configuration, if
+        it is present.
+        """
+        if not self.yarn_path:
+            return None
+
+        matches = _YARNRC_VERSION_REGEX.search(self.yarn_path)
+
+        if matches:
+            return matches[0].replace(".cjs", "").replace("-", "@")
+
+        return None
+
+    @classmethod
+    def from_file(cls, file_path: RootedPath) -> "YarnRc":
+        """Parse the content of a yarnrc file."""
+        with file_path.path.open("r") as file:
+            data = safe_load(file)
+
+        return YarnRc(file_path, data)
+
+
+class PackageJson:
+    """A package.json file.
+
+    This class abstracts the underlying attributes and only exposes what
+    is relevant for the request processing.
+    """
+
+    def __init__(self, path: RootedPath, data: dict[str, Any]) -> None:
+        """Initialize a PackageJson object."""
+        self._path = path
+        self._data = data
+
+    @property
+    def package_manager(self) -> Optional[str]:
+        """Get the package manager string."""
+        return self._data.get("packageManager", None)
+
+    @classmethod
+    def from_file(cls, file_path: RootedPath) -> "PackageJson":
+        """Parse the content of a package.json file."""
+        with file_path.path.open("r") as file:
+            data = json.load(file)
+
+        return PackageJson(file_path, data)
+
+
+class Project(NamedTuple):
```

Comment:
We are not calling a "main input package" a project anywhere else, so I struggled a little with this name here. But I kept it for now, let me know if you have a better suggestion for it.